Connectors/Rhino/Speckle.Connectors.Rhino7/HostApp/RhinoMaterialManager.cs

```diff
@@ -0,0 +1,147 @@
+    }
+
+    using RenderMaterial rm = RenderMaterial.FromMaterial(pbMaterial, null);
+    Rhino.DocObjects.PhysicallyBasedMaterial pbRenderMaterial = rm.ConvertToPhysicallyBased(
+      RenderTexture.TextureGeneration.Allow
+    );
+
+    string renderMaterialName = material.Name ?? "default"; // default rhino material has no name
+    System.Drawing.Color diffuse = pbRenderMaterial.BaseColor.AsSystemColor();
+    System.Drawing.Color emissive = pbRenderMaterial.Emission.AsSystemColor();
+    double opacity = pbRenderMaterial.Opacity;
+
+    SpeckleRenderMaterial speckleRenderMaterial =
+      new(pbRenderMaterial.Opacity, pbRenderMaterial.Metallic, pbRenderMaterial.Roughness, diffuse, emissive)
+      {
+        name = renderMaterialName,
+        applicationId = materialId
+      };
+
+    return speckleRenderMaterial;
+  }
+
+  /// <summary>
+  /// Determines if a Speckle Render Material has already been created from the input Rhino material
+  /// </summary>
+  /// <param name="material"></param>
+  /// <returns>True if yes, False if no Speckle Render material has been created from the input material</returns>
+  public bool Contains(Material material)
+  {
+    return _renderMaterialCache.TryGetValue(material.Id.ToString(), out SpeckleRenderMaterial _);
+  }
+
+  public BakeResult BakeMaterials(
+    List<SpeckleRenderMaterial> speckleRenderMaterials,
+    string baseLayerName,
+    Action<string, double?>? onOperationProgressed
+  )
+  {
+    var doc = RhinoDoc.ActiveDoc; // POC: too much right now to interface around
+
+    // Keeps track of the incoming SpeckleRenderMaterial application Id and the index of the corresponding Rhino Material in the doc material table
+    Dictionary<string, int> materialIdAndIndexMap = new();
+
+    int count = 0;
+    List<ReceiveConversionResult> conversionResults = new();
+    foreach (SpeckleRenderMaterial speckleRenderMaterial in speckleRenderMaterials)
+    {
+      onOperationProgressed?.Invoke("Converting render materials", (double)++count / speckleRenderMaterials.Count);
+      try
+      {
+        // POC: Currently we're relying on the render material name for identification if it's coming from speckle and from which model; could we do something else?
+        string matName = $"{speckleRenderMaterial.name}-({speckleRenderMaterial.applicationId})-{baseLayerName}";
+        Color diffuse = Color.FromArgb(speckleRenderMaterial.diffuse);
+        Color emissive = Color.FromArgb(speckleRenderMaterial.emissive);
+        double transparency = 1 - speckleRenderMaterial.opacity;
+
+        Material rhinoMaterial =
+          new()
+          {
+            Name = matName,
+            DiffuseColor = diffuse,
+            EmissionColor = emissive,
+            Transparency = transparency
+          };
+
+        int matIndex = doc.Materials.Add(rhinoMaterial);
+
+        // POC: check on matIndex -1, means we haven't created anything - this is most likely an recoverable error at this stage
```

Comment:
or does it mean we have it already with the same name? This is also something we need to check if you haven't already. "What will happen if we have the materials with the same name already?"

Reply:
Any material with the same name should've been purged already